### PR TITLE
De-couple Schema Registry from connector, add SchemaRetriever interface

### PR DESCRIPTION
**Pull request by C0urante**
**NOTE**: Package deployment (`gradle uploadArchives`) has yet to be tested.

### CODE REVIEW

**Comment by C0urante** [comment]
Should address https://github.com/wepay/kafka-connect-bigquery/issues/18

**Comment by criccomini** [comment]
Modules should be:

kcbq-api
kcbq-confluent
kcbq-connector

**Comment by criccomini** [comment]
Rather than having a build.gradle in every package, just use a single top-level build.gradle

File: api/gradle.properties [hunk 1]
```diff
@@ -0,0 +1,3 @@
+version = 0.1.0-SNAPSHOT
+group = 'com.wepay.kcbq'
+archivesBaseName = 'kafka-connect-bigquery-schema-retriever'
```
Comment (criccomini) [inline api/gradle.properties]: Update to kcbq-api

File: api/src/main/java/com/wepay/kafka/connect/bigquery/api/SchemaRetriever.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  public void configure(Map<String, String> properties);
+
+  /**
+   * Retrieve the most current schema for the given topic, in order to perform an update in
+   * BigQuery.
+   * @param topic The topic to retrieve a schema for.
+   * @param schemas The set of unique Schemas for the data that caused the insert failure.
+   * @return The Schema for the given table.
+   */
+  public Schema retrieveSchema(String topic, Set<Schema> schemas);
```
Comment (criccomini) [inline api/src/main/java/com/wepay/kafka/connect/bigquery/api/SchemaRetriever.java]: Should table be included here as well? Thinking forward to if we have multiple topics to one table, or one topic to multiple tables?

File: core/gradle.properties [hunk 3]
```diff
@@ -0,0 +1,3 @@
+version = 0.3.0-SNAPSHOT
+group = 'com.wepay.kcbq'
+archivesBaseName = 'kafka-connect-bigquery-connector'
```
Comment (criccomini) [inline core/gradle.properties]: kcbq-connector

File: core/src/integration-test/resources/test.properties [hunk 4]
```diff
@@ -0,0 +1,3 @@
+keyfile=/Users/chrise/Work/kafka-connect-bigquery/key.json
```
Comment (criccomini) [inline core/src/integration-test/resources/test.properties]: delete this proprietary stuff

File: core/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java [hunk 5]
```diff
@@ -232,14 +250,28 @@ private void pauseAllPartitions(String topic) {
...
         : partitionRecordsByTable(records).entrySet()) {
-      TableId table = tableRow.getKey();
+      TableId table = tableRecords.getKey();
+
       if (!tableBuffers.containsKey(table)) {
         tableBuffers.put(table, getNewBuffer());
       }
+
+      if (!tableSchemas.containsKey(table)) {
+        tableSchemas.put(table, new HashSet<>());
```
Comment (criccomini) [inline core/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java]: `Collections.emptySet();`

File: core/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java [hunk 6]
```diff
@@ -69,20 +75,21 @@
       "Names for the datasets kafka topics will write to "
       + "(form of <topic regex>=<dataset>)";
 
+  public static final String SCHEMA_RETRIEVER_CONFIG =         "schemaRetriever";
+  private static final ConfigDef.Type SCHEMA_RETRIEVER_TYPE =  ConfigDef.Type.CLASS;
+  private static final Class<?> SCHEMA_RETRIEVER_DEFAULT =     null;
```
Comment (criccomini) [inline core/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java]: Being lazy: what happens if this is null? Do we default to the confluent one for backwards compatibility?

File: core/test/docker/connect/properties/connector.properties [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+sanitizeTopics=true
+
+bufferSize=100000
+maxWriteSize=1000
+tableWriteWait=1000
+
+keyfile=/tmp/key.json
+schemaRetriever=com.wepay.kafka.connect.bigquery.schemaregistry.schemaretriever.SchemaRegistrySchemaRetriever
+schemaRegistryLocation=http://schema-registry:8081
+project=steady-run-1337
```
Comment (criccomini) [inline core/test/docker/connect/properties/connector.properties]: Clear proprietary stuff.

File: schema-registry-schema-retriever/gradle.properties [hunk 8]
```diff
@@ -0,0 +1,3 @@
+version = 0.1.0-SNAPSHOT
+group = 'com.wepay.kcbq'
+archivesBaseName = 'kafka-connect-bigquery-schema-registry-retriever'
```
Comment (criccomini) [inline schema-registry-schema-retriever/gradle.properties]: kcbq-kafka

File: core/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java (hunk 6 above)
Comment (C0urante) [inline core/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java]: Right now, leaving `schemaRetriever` null just causes automatic table creation / schema updates to be disabled (and an exception thrown if the config specifies that they should be enabled).

File: schema-registry-schema-retriever/gradle.properties (hunk 8 above)
Comment (C0urante) [inline schema-registry-schema-retriever/gradle.properties]: kcbq-confluent?

File: core/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java (hunk 6 above)
Comment (C0urante) [inline core/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java]: Thinking of just adding another Gradle task (something like `confluentTarBall`) in the kcbq-connector directory that just combines the connector and confluent tarballs.

Comment (criccomini) [inline core/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java]: Ah. Got it. OK, default should throw exception then, as you have it currently.

File: schema-registry-schema-retriever/gradle.properties (hunk 8 above)
Comment (criccomini) [inline schema-registry-schema-retriever/gradle.properties]: Sorry, yea. kcbq-confluent

**Pull request merged**